lib/processor.js

```diff
@@ -0,0 +1,111 @@
+'use strict'
+
+const contentful = require('contentful')
+const validator = require('./validator')
+const util = require('./util')
+
+/**
+ * Create contentful client
+ *
+ * @param  {String} accessToken access token
+ * @param  {String} spaceId     space id
+ *
+ * @return {Object}             contentful client
+ */
+function createContentfulClient (accessToken, spaceId, host) {
+  return contentful.createClient({
+    space: spaceId,
+    accessToken,
+    host
+  })
+}
+
+/**
+ * Enrich all fetched entries with additional properties
+ *
+ * @param {Array}  entries fetched entries
+ * @param {Object} file    file the entries were fetched for
+ *
+ * @return {Array}         enriched entries
+ */
+function mapEntriesForFile (entries, file, options) {
+  return entries.map(entry => {
+    entry._fileName = util.getFileName(entry, file.contentful, options)
+
+    return entry
+  })
+}
+
+/**
+ * Process the fetched entries by contentful
+ * for given file
+ *
+ * @param {Object} file    file read by metalsmith
+ * @param {Array}  entries entries fetched from contentful
+ *
+ * @return {Object}        file mapping object
+ */
+function processEntriesForFile (file, entries) {
+  const options = file.contentful
+  const files = {}
+
+  files[file._fileName] = file
+
+  if (options.entry_id) {
+    validator.validateSingleEntryForFile(entries[0], file)
+
+    file.data = entries[0]
+  } else {
+    file.data = { entries }
+  }
+
+  if (options.entry_template) {
+    return entries.reduce((fileMap, entry) => {
+      fileMap[ entry._fileName ] = {
+        // `contents` need to be defined because there
+        // might be other plugins that expect it
+        contents: '',
+        data: entry,
+        id: entry.sys.id,
+        contentType: options.content_type,
+        layout: options.entry_template,
+        _fileName: entry._fileName
+      }
+
+      return fileMap
+    }, files)
+  }
+
+  return files
+}
+
+/**
+ * Process one file and connect it with contentful data
+ *
+ * @param {Object} file    file read by metalsmith
+ * @param {Object} options contentful metalsmith options
+ *
+ * @return {Object|Promise}
+ */
+function processFile (file, options) {
+  if (!file.contentful) {
+    return true
+  }
+
+  validator.validateFile(file)
+  validator.validateFileAndOptions(file, options)
+
+  const spaceId = file.contentful.space_id || options.space_id
+  const host = file.contentful.host || options.host
+  const query = util.getEntriesQuery(file.contentful)
+
+  const client = createContentfulClient(options.access_token, spaceId, host)
```

Comment:
I think a shared client for all files would be better or the contentful internal rate-limiting logic and should also simplify testing as proxyquire would no longer be required.